c10/cuda/CUDAFunctions.cpp

```diff
@@ -182,4 +182,40 @@ bool hasPrimaryContext(int64_t device_index) {
   return _internal::hasPrimaryContext(device_index);
 }
 
+std::atomic_int LastSavedDevice = -1;
+
+// Wrappers for raw CUDA device management functions
+cudaError_t GetDeviceCount(int* dev_count) {
+  return cudaGetDeviceCount(dev_count);
+}
+
+cudaError_t GetDevice(int* device) {
+  if (LastSavedDevice >= 0) {
+    *device = LastSavedDevice;
+    return cudaSuccess;
+  }
+  return cudaGetDevice(device);
+}
+
+cudaError_t SetDevice(int device) {
+  TORCH_CHECK(device >= 0, "device id must be positive!");
+  LastSavedDevice = -1;
```

Comment:
do we need atomic write for every SetDevice? Device is thread local in any case, so why do we need atomic here?

Reply:
For some reason RPC tests went flaky and we figured out that it was trying to call set device from different threads ending up in race conditions. 
I shall try using thread local. However, one of my attempts failed: https://github.com/pytorch/pytorch/actions/runs/4495850640/jobs/7909930986